File: src/services/tools/code/findFiles.ts
import { z } from 'zod';
import type { ToolDef } from '../types';
import { isTauri } from '../../../lib/tauri';

async function loadFs() {
  if (!isTauri()) return null;
  try {
    return await import('@tauri-apps/plugin-fs');
  } catch {
    return null;
  }
}

function matchesGlob(name: string, pattern: string): boolean {
  const regex = new RegExp('^' + pattern.replace(/\*/g, '.*').replace(/\?/g, '.') + '$', 'i');
  return regex.test(name);
}

export const findFilesTool: ToolDef = {
  name: 'find_files',
  description: 'Recursively find files matching a glob pattern within a directory. Returns matching file paths.',
  category: 'code',
  inputSchema: z.object({
    pattern: z.string().describe('Glob pattern to match file names against (e.g. "*.ts", "*.{ts,tsx}", "**/*.test.*").'),
    path: z.string().describe('Absolute path to the root directory to search in.'),
    maxResults: z.number().optional().default(100).describe('Maximum number of results to return.'),
  }),
  execute: async ({ pattern, path, maxResults }) => {
    const mod = await loadFs();
    if (!mod) throw new Error('Filesystem access requires the Tauri desktop app');
    const fs = mod;

    const results: string[] = [];

    async function walk(dir: string): Promise<void> {
      if (results.length >= maxResults!) return;
      try {
        const entries = await fs.readDir(dir);
        for (const entry of entries) {
          if (results.length >= maxResults!) return;
          const fullPath = `${dir}/${entry.name}`;
          if (entry.isDirectory) {
            await walk(fullPath);
          } else if (entry.isFile && matchesGlob(entry.name, pattern)) {
            results.push(fullPath);
          }
        }
      } catch {
        // skip directories we cannot read
      }
    }

    await walk(path);
    return { path, pattern, results, count: results.length, truncated: results.length >= maxResults! };
  },
};

File: agent/internal/tool/find_files.go
package tool

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func findFilesTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "find_files",
			Description: "Find files by name or glob pattern recursively from a root directory.",
			Category:    "code",
			Parameters: map[string]api.ParamDef{
				"pattern":     {Type: "string", Description: "File name or glob pattern to search for (e.g. '*.ts', 'main.*', '**/test*')", Required: true},
				"root":        {Type: "string", Description: "Root directory to start search from", Required: false, Default: "."},
				"max_results": {Type: "number", Description: "Maximum files to return", Required: false, Default: 50},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			pattern, _ := params["pattern"].(string)
			if pattern == "" {
				return nil, fmt.Errorf("pattern is required")
			}
			root, _ := params["root"].(string)
			if root == "" {
				root = "."
			}
			root = expandPath(root)
			maxResults, _ := params["max_results"].(float64)
			if maxResults == 0 {
				maxResults = 50
			}

			var matches []string
			filepath.Walk(root, func(fp string, info os.FileInfo, err error) error {
				if err != nil {
					return nil
				}
				if len(matches) >= int(maxResults) {
					return filepath.SkipAll
				}
				matched, _ := filepath.Match(pattern, info.Name())
				if matched {
					matches = append(matches, fp)
				}
				return nil
			})

			return map[string]any{
				"pattern": pattern,
				"root":    root,
				"matches": matches,
				"count":   len(matches),
			}, nil
		},
	}
}
